File: libraries/bicycle_model/src/bicycle_model_calculations.cpp

#include "bicycle_model/bicycle_model_calculations.h"

geometry_msgs::Vector3 changePosition(double linearVel, double steerAngle, double timeSecs, double carLength) {
  return forwardKinematics(linearVel, angularVelocity(linearVel, steerAngle, carLength), timeSecs);
}


geometry_msgs::Vector3 forwardKinematics(double linearVel, double angularVel, double timeSecs) {
  double dth = angularVel * timeSecs;
  double dx  = linearVel * sin(dth) / angularVel;
  double dy  = linearVel * (1 - cos(dth)) / angularVel;

  geometry_msgs::Vector3 msg;
  msg.x = dx;
  msg.y = dy;
  msg.z = dth;
  return msg;
}

double angularVelocity(double linearVel, double steerAngle, double carLength) {
  return linearVel * tan(steerAngle) / carLength;
}

File: libraries/bicycle_model/scripts/BicycleModelCalculations.py
import math

""" Calculates the change in position in the robot frame
    v: forward velocity of the car
    angle: steering angle of the car
    time: time lapse for motion """
def changePosition(v, angle, time, carLength):
    return forwardKinematics(v, angularVelocity(v, angle, carLength), time)

""" Calculates the change in position in the robot frame
    v: forward velocity of the car
    w: angular velocity of the car
    time: time lapse for motion """
def forwardKinematics(v, w, time):
    dx = v * math.sin(w * time) / w
    dy = v * (1 - math.cos(w * time)) / w
    dtheta = w * time
    return (dx, dy, dtheta)

""" Calculates angular velocity from car inputs
    v: forward velocity of the car
    angle: steering angle of the car """
def angularVelocity(v, angle, carLength):
    return v * math.tan(angle) / carLength

""" Converts a points in the world frame (xW, yW, thetaW) to
    a point in the robot frame given the robot position in
    the world (carX, carY, carTheta) """
def toRobotFrame(xW, yW, thetaW, carX, carY, carTheta):
    xR = (xW - carX) * math.cos(carTheta) + (yW - carY) * math.sin(carTheta)
    yR = (carX - xW) * math.sin(carTheta) + (yW - carY) * math.cos(carTheta)
    thetaR = thetaW - carTheta
    return (xR, yR, thetaR)

""" Converts a points in the robot frame (xR, yR, thetaR) to
    a point in the world frame given the robot position in
    the world (carX, carY, carTheta) """
def toWorldFrame(xR, yR, thetaR, carX, carY, carTheta):
    xW = xR * math.cos(carTheta) - yR * math.sin(carTheta) + carX
    yW = xR * math.sin(carTheta) + yR * math.cos(carTheta) + carY
    thetaW = (thetaR + carTheta) % (2 * math.pi)
    return (xW, yW, thetaW)